docs/testing-after-commit-callbacks.md

```diff
@@ -0,0 +1,104 @@
+# Testing after-commit callbacks
+
+## The Problem
+
+Django's after-commit callbacks don't work properly in tests when using `django.db.transaction.atomic`.
+This creates a disconnect between test behaviour and production behaviour, potentially hiding bugs.
+
+Consider this function that should print `A`, `B`, `C`, `D` in order:
+
+```python
+from functools import partial
+from django.db import transaction
+
+def build_ABCD():
+    my_list = []
+    with transaction.atomic():
+        my_list.append("A")
+        transaction.on_commit(partial(my_list.append, "C"))
+        my_list.append("B")
+    my_list.append("D")
+    return my_list
+```
+
+### In production
+
+This returns `["A", "B", "C", "D"]` as expected.
+
+### In tests
+
+```python
+from django.test import TestCase
+
+class TestBuildABCD(TestCase)
+    def test_build_ABCD():
+        assert build_ABCD() == ["A", "B", "C", "D"] # This will fail!
+```
+
+This fails because `build_ABCD()` returns `["A", "B", "D"]`,
+due to the fact that the after-commit callback never runs!
+
+## Why this happens
+
+Django's `TestCase` runs each test in a transaction,
+which is rolled back at the end of the test to prevent test pollution.
+Because the test transaction never commits, Django does not run the after-commit callbacks.
+
+## The Solution: `django_subatomic.db.transaction`
+
+Use `django_subatomic.db.transaction` instead of `django.db.transaction.atomic` in your application code
+
+```diff
+ from functools import partial
++from django_subatomic import db
+ from django.db import transaction as django_transaction
+
+ def build_ABCD():
+     my_list = []
+-    with transaction.atomic():
++    with db.transaction():
+         my_list.append("A")
+         django_transaction.on_commit(partial(my_list.append, "C"))
+         my_list.append("B")
+     my_list.append("D")
+     return my_list
+
+```
+
+`django_subatomic.db.transaction` explicitly represents a transaction,
+so tests can safely run after-commit callbacks when it exits.
+This provides realistic production behaviour without the downsides of other approaches.
+
+## Alternatives considered
+
+Before developing `django-subatomic`, two alternative approaches were considered
+to mitigate the above problem of Django's after-commit callbacks not working properly.
+
+### ⚠️ Using pytest-django `django_capture_on_commit_callbacks` fixture (timing issues)
+
+```python
+
+def test_with_capture(django_capture_on_commit_callbacks):
+    with django_capture_on_commit_callbacks():
+            build_ABCD()
+```
```

Comment:
Issue: We should write these tests with `TestCase`, not Pytest:

```python
from django.test import TestCase

class TestBuildABCD(TestCase):
    def test_build_ABCD():
        with self.captureOnCommitCallbacks(execute=True)
            built = build_ABCD()
        assert built == ["A", "B", "C", "D"]  # This will fail!
```